cmd/gitserver/server/vcs_syncer_npm_packages.go

```diff
@@ -365,109 +375,40 @@ func decompressTgz(tgzReadSeeker namedReadSeeker, destination string) (err error
 	// down further. So we strip the superfluous directory if we detect one.
 	//
 	// https://github.com/sourcegraph/sourcegraph/pull/28057#issuecomment-987890718
-	var superfluousDirName *string = nil
-	commonSuperfluousDirectory := true
-	err = withTgz(tgzReadSeeker, func(reader *tar.Reader) (err error) {
-		for {
-			header, err := reader.Next()
-			if err == io.EOF {
-				return nil
-			}
-			if err != nil {
-				return errors.Wrapf(err, "failed to read tar file %s", tgzReadSeeker.name)
-			}
-			switch header.Typeflag {
-			case tar.TypeReg:
-				components := strings.SplitN(header.Name, string(os.PathSeparator), 2)
-				if len(components) != 2 {
-					commonSuperfluousDirectory = false
-					return nil
-				}
-				outermostDir := components[0]
-				if superfluousDirName == nil {
-					superfluousDirName = &outermostDir
-				} else if *superfluousDirName != outermostDir {
-					commonSuperfluousDirectory = false
-					return nil
-				}
-			default:
-				continue
-			}
-		}
-	})
+
+	files, err := os.ReadDir(dir)
 	if err != nil {
 		return err
 	}
-	if _, err := tgzReadSeeker.value.Seek(0, io.SeekStart); err != nil {
-		return err
-	}
-	if !commonSuperfluousDirectory {
-		log15.Warn("found npm tarball which doesn't have all files in one top-level directory", "label", tgzReadSeeker.name)
-	}
-	// Post-condition: if commonSuperfluousDirectory is true
-	// then for all files in the tarball, the path (in header.Name)
-	// has at least one path separator.
 
-	return withTgz(tgzReadSeeker, func(tarReader *tar.Reader) (err error) {
-		destinationDir := strings.TrimSuffix(destination, string(os.PathSeparator)) + string(os.PathSeparator)
-		for {
-			header, err := tarReader.Next()
-			if err == io.EOF {
-				return nil
-			}
-			switch header.Typeflag {
-			case tar.TypeDir:
-				continue // We will create directories later; empty directories don't matter for git.
-			case tar.TypeReg:
-				name := header.Name
-				if commonSuperfluousDirectory {
-					name = strings.SplitN(name, string(os.PathSeparator), 2)[1]
-				}
-				cleanedOutputPath, isPotentiallyMalicious := isPotentiallyMaliciousFilepathInArchive(name, destinationDir)
-				if isPotentiallyMalicious {
-					continue
-				}
-				err = copyTarFileEntry(header, tarReader, cleanedOutputPath)
-				if err != nil {
-					return err
-				}
-			default:
-				return errors.Errorf("unrecognized type of header %+v in tarball for %s", header.Typeflag, tgzReadSeeker.name)
-			}
+	outermostDir := dir
+	for _, f := range files {
+		if !f.IsDir() || f.Name() == ".git" {
```

Comment:
```suggestion
	for _, dirEntry := range dirEntries {
	    if dirEntry.IsDir() && dirEntry.Name() == ".git" {
			continue
		}
		if !dirEntry.IsDir() { // can || this with the outermostDir != dir check below
		    return nil
		}
```

IIUC, with the `continue` instead of the `return nil`, if faced with a directory layout like:

```
dir
 +------ file1
  |------ .git/
  \------ dir2/
```

`file1` will get dropped because only `dir2/` and `.git/` will be copied.

Reply:
https://github.com/sourcegraph/sourcegraph/pull/32320/commits/face6981fa3e0e8b8fdd781c9aafc899a109e8c6